drivers/hmis/app/models/hmis/form/instance.rb

```diff
@@ -10,7 +10,9 @@ class Hmis::Form::Instance < ::GrdaWarehouseBase
   self.table_name = :hmis_form_instances
 
   belongs_to :entity, polymorphic: true, optional: true
-  belongs_to :definition, foreign_key: :definition_identifier, primary_key: :identifier, class_name: 'Hmis::Form::Definition'
+  belongs_to :definition,
+             -> { order(Arel.sql("CASE WHEN status = 'published' THEN 0 WHEN status = 'draft' THEN 1 ELSE 2 END")) },
+             foreign_key: :definition_identifier, primary_key: :identifier, class_name: 'Hmis::Form::Definition'
```

Comment:
Is there any reason we shouldn't just change this to a `has_many`, since the instance may have multiple definition records? That feels clearer to me. I am only finding 2 usages of this:
* DeleteFormRule (can just use `definitions.first`)
* UpdateFormRule (deprecated mutation, but same deal)

Reply:
Hm, yeah I agree the `has_many` relationship makes a lot more sense. But I think there might be a few more usages we need to check, like in `Project.available_service_types`, scopes internal to `Instance` and also `Definition`, `enabled_features` on client. For each of those places where we are joining from Instance to Definition, if we change from `belongs_to` to `has_many`, we have to make sure that the scope doesn't then contain duplicate Instances, correct?

Reply:
I'll give it a go for sure, but I'm also wondering if now is the time to separate out the two unrelated bugfixes in this PR so the case notes form definition ID fix isn't blocked by this.

Reply:
Ahh okay, yes. I didn't think about the places where we join instances on the `:definition` relation. I think your approach is probably best (and least disruptive), then.

Reply:
Thanks 🙏 I'll add some comments

Reply:
And, I'm sorry, I feel like I had this but then I lost it. How does this specific change affect the Client Details or Case Notes bug?

Reply:
Yes so, without this change, the `definition` on the Instance points to the first definition with this identifier, which isn't necessarily the published one (could be draft or, more likely, retired). `HmisSchema::OccurrencePointForm` then gets returned to the frontend with a `definition` that is not the currently published version, and that's the definition used for pathway editing.

You can see the bug in action by reverting the change locally and then running `client_detail_forms_spec`.

Reply:
Okay I see, thanks.  I think the desired state is that `project { occurrencePointForms }` and `clientDetailForms` should _only_ resolve forms that are published. (There is no use-case for "retired" on those, from what I can think up right now).

So this does seem to accomplish that, because you added the `published` scope, and updated this `definition` relation to _prefer_ the publish form, which results in OccurrencePointForm's `load_ar_association(object, :definition)` loading the correct form. Am I following that correctly?

I wonder if we can make things a bit clearer & safer in the OccurrencePointForm object resolver. We know that we only ever want to resolve a published FormDefinition here, so we can enforce that:

```ruby
definition = load_ar_association(object, :definitions).select(&:published?) # uses has_many relation

# or

definition = load_ar_association(object, :published_definition) # uses belongs_to :published_definition, -> { published }

# or, with your sorted relation

definition = load_ar_association(object, :definition) # this is sorted to prefer published definitions
raise 'bad' unless definition.published?
```


Does that make sense?

Reply:
@gigxz yes your summary of the bug/fix matches my understanding, and yes what you wrote does make sense. I think I prefer one of your first two suggestions, even though either would add some complexity with an extra relation on the Instance model. An admin hypothetically could add an Instance on a Client Detail form that is still a Draft, and then option 3 would cause an uncaught exception, whereas option 1 or 2 would simply exclude that form from the list, I think.

Reply:
All 3 could potentially raise if we forget to filter to published instances only, I updated my example. I guess I lean towards adding the `has_many :definitions` relation and going that route, since it is descriptive of the actual db relationship. I don't see that as extra complexity, since it is the more accurate relation. We may want to use that one going forward for other uses. (In addition to your change in the `belongs_to :definition` which I think is wise. That relation feels problematic in general, but at least its better if its preferring the published one..).

Here I was thinking we crossed all our Ts with the form publishing lifecycle 😅

Reply:
Now I see you added published_definition -also happy to go with that!

Reply:
Ah yeah it's a good point about the has many relationship. We should create it now and gradually transition to using it instead of the `belongs_to :definition`.